dev-docs/design/ssr-react.md

```diff
@@ -0,0 +1,96 @@
+
+#### Patching `createElement`
+
+The classic JSX transform compiles JSX to `React.createElement` calls. Our `@lit-labs/react` wrapper also uses the `React` passed in by user to call `createElement` on it.
+
+We can provided an alternate `createElement` function that applies the core strategy above.
+
+For convenience, we can provide a side-effectful import that monkey patches `React.createElement` to be this alternate version.
+
+Using export conditions, we can create a module `@lit-labs/ssr-react/enable-lit-ssr.js` that when imported in Node will monkey patch `React.createElement` and when imported in the browser will load the `@lit-labs/ssr-client/lit-element-hydrate-support.js`.
+
+#### Using runtime JSX
+
+The runtime JSX transform compiles JSX to various jsx function calls, importing `jsx` or `jsxs` from `<package>/jsx-runtime` or `jsxDEV` from `<package>/jsx-dev-runtime` depending on the environment and whether the `children` array was created statically.
+
+We can provide alternate jsx functions that enhance the base React jsx functions with the core strategy above.
+
+This is mainly for targeting bare custom elements as wrapped components must have `React.createElement` patched to work.
+
+### `@lit-labs/react` package
+
+The wrapper made with `createComponent` from the Node build of this package will check whether `React.createElement` has been patched by checking its `.name` patches ours. If so, it will pass element properties in a special prop named `_$litProps$` to `createElement` to be used for server rendering as described above.
+
+On the client, the wrapper on mount will set properties on the element and remove the `defer-hydration` attribute to start Lit element hydration. It will also add `suppressHydrationWarning` to the prop of the host element to suppress React's warning for having extra attributes on the server rendered HTML like `defer-hydration`.
+
+There was consideration of baking in the whole SSR template shadowroot generation into the Node build of `@lit-labs/react` package itself which would remove the need for monkey patching `createElement`, but users would then have to make sure to manually load the `lit-element-hydrate-support` early anyway. It's not out of the realms of possiblity.
+
+### `@lit-labs/nextjs` package
+
+This is a plugin for Next.js projects to wrap the config in `next.config.js`, adding custom ebpack config that automatically imports `@lit-labs/ssr-react/enable-lit-ssr.js` and replaces the `react/jsx-runtime` imports with `@lit-labs/ssr-react/jsx-runtime`.
+
+While users are able to specify a `jsxImportSource` for the runtime JSX, this won't work for any external components that were pre-compiled with runtime JSX, nor does it allow composing multiple runtime jsx functions. Replacing the module at the webpack config level addresses these.
```

Comment:
This paragraph may need some additional clarification. I'm not understanding how it fits in with this doc. The doc has been very detailed regarding the property/attribute handling, as well as differences between bare custom elements vs labs/React wrapped elements. For me this section is missing background.

Why would I specify a `jsxImportSource`, and what is an external component? What is the behavior in nextjs without the labs package?

Maybe remove from this PR, as it could be its own `design/nextjs.md` doc.

A slight meta point. This document is called `ssr-react.md` which suggests to me it's the canonical design for React ssr. However after reading it may be better to focus it on this complex feature of property & attribute handling with React SSR.

This is a great doc! Thanks for writing it up!

Reply:
I did want to include everything in our ecosystem related to React here and not have a design doc per labs package kind of deal as they end up working together.

The point about `jsxImportSource` maybe is better explained here: https://github.com/lit/lit/tree/main/packages/labs/ssr-react#using-the-automatic-runtime-jsx-transform

External component refers to components you bring in from outside the project like npm or design system library.

Behavior of Next.js without labs package is as described above as it's just React SSR.

Thank you for the feedback! I'll add more context/clarify these points.